core/trino-main/src/main/java/io/trino/operator/project/PageFieldsToInputParametersRewriter.java

```diff
@@ -42,80 +45,117 @@ private PageFieldsToInputParametersRewriter() {}
     public static Result rewritePageFieldsToInputParameters(RowExpression expression)
     {
         Visitor visitor = new Visitor();
-        RowExpression rewrittenProjection = expression.accept(visitor, null);
-        InputChannels inputChannels = new InputChannels(visitor.getInputChannels());
+        RowExpression rewrittenProjection = expression.accept(visitor, true);
+        InputChannels inputChannels = new InputChannels(visitor.getInputChannels(), visitor.getEagerlyLoadedChannelIndexes());
         return new Result(rewrittenProjection, inputChannels);
     }
 
     private static class Visitor
-            implements RowExpressionVisitor<RowExpression, Void>
+            implements RowExpressionVisitor<RowExpression, Boolean>
     {
         private final Map<Integer, Integer> fieldToParameter = new HashMap<>();
         private final List<Integer> inputChannels = new ArrayList<>();
+        private final Set<Integer> eagerlyLoadedChannels = new HashSet<>();
         private int nextParameter;
 
         public List<Integer> getInputChannels()
         {
             return ImmutableList.copyOf(inputChannels);
         }
 
+        public List<Integer> getEagerlyLoadedChannelIndexes()
+        {
+            return IntStream.range(0, inputChannels.size()).boxed()
```

Comment:
I'm returning indices of eagerly loaded channels in `inputChannels` rather than the eagerly loaded channels themselves so that in `Page#getLoadedPage` we don't need to perform any look-ups.